[core] Split raylet cython file into multiple files (ObjectRefGenerator)#56520

Artimislyy opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
As mentioned in https://github.com/ray-project/ray/issues/51080, separate ObjectRefGenerator class from the large _raylet.pyx file.
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
close https://github.com/ray-project/ray/issues/56520

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Moves `ObjectRefGenerator` out of `ray/_raylet.pyx` into `ray/_private/object_ref_generator.py` and updates imports/tests accordingly.
> 
> - **Core**:
>   - **Module split**: Move `ObjectRefGenerator` implementation from `ray/_raylet.pyx` to `ray/_private/object_ref_generator.py` (keeping API: sync/async iteration, `completed()`, `next_ready()`, `is_finished()`).
>   - **Imports**: `ray/_raylet.pyx` now imports `ObjectRefGenerator, DynamicObjectRefGenerator` from `ray/_private/object_ref_generator.py`; removes inlined class.
>   - **Cleanup**: Ensure stream deletion in `__del__` via `core_worker.async_delete_object_ref_stream` and disallow serialization via `__getstate__`.
> - **Tests**:
>   - Update `test_streaming_generator.py` to explicitly call `generator.__del__()` before `del` and assert stream deletion call.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 74e8d41fa1ca3283ff7a570376e286011498ab6e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
split ObjectRefGenerator

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Merge branch 'master' into ObjectRefGenerator

edoakes reviewed (commented):
I just commented on the issue -- to be safe, let's just keep both class definitions in the same `object_ref_generator.py` file.

Artimislyy committed:
Merge branch 'ray-project:master' into ObjectRefGenerator

Commit:
merge two files #56575

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Artimislyy committed:
Merge branch 'ray-project:master' into ObjectRefGenerator

Artimislyy commented on python/ray/_private/object_ref_generator.py:
```diff
@@ -23,3 +23,250 @@ def __iter__(self) -> Iterator("ray.ObjectRef"):
+
+@PublicAPI
+class ObjectRefGenerator:
+    """A generator to obtain object references
```
Yes, I copied and pasted it, but running `scripts/format.sh` caused the formatting to change.

Commit:
Change the file format

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Merge branch 'master' into ObjectRefGenerator

Commit:
Merge branch 'ObjectRefGenerator' of github.com:Artimislyy/ray into ObjectRefGenerator

Commit:
Delete the extra import line

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

edoakes commented:
@Artimislyy there are some CI failures: https://buildkite.com/ray-project/microcheck/builds/26178#01996024-a15d-477b-8511-dab69f1f0195/7-5294

Commit:
change ObjectRef to ray.ObjectRef

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

jjyao commented:
CI is failing

Commit:
add worker

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
add TYPE_CHECKING

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the comments of _next_sync and resolve the circular import issue.

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
solve  Ray Module Not Loaded at Runtime

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the file format of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Modify the comments for next(gen) and gen.completed()

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Artimislyy commented:
> CI is failing
Hi @jjyao.
I copied the original code including the comments, but the CI is showing errors in them. 
For example
```
>>> gen = generator_task.remote()
>>> next(gen)
>>> await gen.__anext__()
```
the CI indicates that `generator_task` is not defined. Additionally, the comment for the `_next_sync` function is missing the return section.
I've made some changes to pass the CI checks, but I'm not sure if they fully meet your requirements.I'd like to know why this happened.

Commit:
Modify the comments for await gen.__anext__()

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
modify test_streaming_generator.py and the comments of object_ref_generator.py

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
delete c.async_delete_object_ref_stream.reset_mock()

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Artimislyy committed:
Merge branch 'master' into ObjectRefGenerator

Commit:
add ray.init()

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Commit:
Merge branch 'ObjectRefGenerator' of github.com:Artimislyy/ray into ObjectRefGenerator

Artimislyy committed:
Merge branch 'master' into ObjectRefGenerator

Commit:
modify ray.init()

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Artimislyy committed:
Merge branch 'master' into ObjectRefGenerator

Commit:
delect the comment of generator_task

Signed-off-by: l00951262 <luoyangyang12@huawei.com>

Artimislyy committed:
Merge branch 'master' into ObjectRefGenerator

Artimislyy commented:
@edoakes CI premerge tests have passed.

edoakes commented on python/ray/tests/test_streaming_generator.py:
```diff
@@ -93,6 +93,7 @@ def test_streaming_object_ref_generator_basic_unit(mocked_worker):
             with pytest.raises(TypeError):
                 dumps(generator)
 
+            generator.__del__()
```
hm, why is this line necessary? it should be called on the subsequent line unless there is a circular reference in the test (or worse, in the async generator code itself)

Artimislyy replied:
> hm, why is this line necessary? it should be called on the subsequent line unless there is a circular reference in the test (or worse, in the async generator code itself)
```
with pytest.raises(StopIteration):
       ref = generator._next_sync(timeout_s=0)
```
When pytest.raises catches a StopIteration, it keeps the exception object alive; the traceback holds a reference to the generator, so the generator’s __del__ is never called.

Artimislyy committed:
Merge branch 'ray-project:master' into ObjectRefGenerator

edoakes reviewed (changes requested):
Still concerned about the need to manually call `__del__` as it might indicate a reference cycle having been introduced. Otherwise LGTM

edoakes committed:
Merge branch 'master' of https://github.com/ray-project/ray into ObjectRefGenerator

edoakes committed:
fix merge

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
remove __del__

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

Artimislyy commented:
> I pushed a commit to resolve the merge conflicts and also cleaned up a few minor things:
> 
> 1. Storing a boolean instead of the full task exception.
> 2. Cleaning up formatting on the docstring.
> 
> I also removed the manual `__del__` call from the test and ran it 10000 times locally without failure. Did it fail locally for you without the call?



> I pushed a commit to resolve the merge conflicts and also cleaned up a few minor things:
> 
> 1. Storing a boolean instead of the full task exception.
> 2. Cleaning up formatting on the docstring.
> 
> I also removed the manual `__del__` call from the test and ran it 10000 times locally without failure. Did it fail locally for you without the call?

During my local testing, I indeed encountered the issue. I'm not sure why you didn't experience this problem, but I confirmed that the CI system actually has this issue. The Python test [g5_s5] problem is `AssertionError: Expected 'async_delete_object_ref_stream' to have been called`, which matches the issue I encountered. This is why I added the` __del__()` method. Additionally, I previously encountered The doc tests [g5_s12] show: `EXCEPTION: NameError("name 'gen' is not defined")`. Therefore, I modified some code comments to pass the CI checks.

edoakes committed:
testcode

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
testcode

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
testcode

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
add gc.collect

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
add gc.collect

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
Merge branch 'master' of https://github.com/ray-project/ray into ObjectRefGenerator

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
Merge branch 'master' of https://github.com/ray-project/ray into ObjectRefGenerator

edoakes committed:
try raising from None

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>

edoakes committed:
fix

Signed-off-by: Edward Oakes <ed.nmi.oakes@gmail.com>
